trieste/space.py

```diff
@@ -331,3 +331,140 @@ def __mul__(self, other: Box) -> Box:
+
+
+class OrdinalSearchSpace(Box):
+    r"""
+    Ordinal search space representing a discrete :math:`D`-dimensional box in
+    :math:`\mathbb{R}^D`. This class inherit :class:`Box` and generates evenly-spaced value
+    in an interval for each dimension. For example randomly sample from this search space
+    will give:
+
+        >>> tf.random.set_seed(1945)
+        >>> ordinalsp = OrdinalSearchSpace([-2.0, -2.0], [2.0, 2.0], [0.2, 0.3])
+        >>> sample = ordinalsp.sample(5)
+        >>> sample.numpy()
+        array([[ 0. , -0.6],
+               [ 0.6, -0.6],
+               [-1.2,  1.2],
+               [-0.4,  1.5],
+               [ 1.2,  0. ]])
+
+    """
+
+    def __init__(
+        self,
+        lower: Sequence[float] | TensorType,
+        upper: Sequence[float] | TensorType,
+        stepsizes: Sequence[float] | TensorType,
+    ):
+
+        r"""
+        If ``lower``, ``upper`` and ``stepsize`` are `Sequence`\ s of floats
+        (such as lists or tuples), they will be converted to tensors of dtype `tf.float64`.
+
+        :param lower: The lower (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param upper: The upper (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param stepsizes: The step size of each dimension of the Ordinal Space. Must have
+            shape [D] for positive D, and if a tensor, must have float type.
+        :raise ValueError (or InvalidArgumentError): If any of the following are true:
+
+            - ``lower``, ``upper`` and ``stepsizes`` have invalid shapes.
+            - ``lower``, ``upper`` and ``stepsizes`` do not have the same floating point type.
+            - ``upper`` is not greater than ``lower`` across all dimensions.
+        """
+
+        super().__init__(lower, upper)
+        tf.debugging.assert_shapes([(lower, ["D"]), (stepsizes, ["D"])])
+        tf.assert_rank(stepsizes, 1)
+
+        tf.debugging.assert_positive(len(stepsizes), message="stepsizes cannot be empty")
+
+        if isinstance(stepsizes, Sequence):
+            self._stepsizes = tf.constant(stepsizes, dtype=tf.float64)
+        else:
+            self._stepsizes = tf.convert_to_tensor(stepsizes)
+
+            tf.debugging.assert_same_float_dtype([self._lower, self._stepsizes])
+
+    def __repr__(self) -> str:
+        """"""
+        return f"OrdinalSearchSpace({self._lower!r}, {self._upper!r}, {self._stepsizes!r})"
+
+    def sample(self, num_samples: int) -> TensorType:
+        """
+        :param num_samples: The number of points to sample from this search space.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly, and without replacement,
+            from this search space.
+        """
+        samples = super().sample(num_samples)
+        samples = tf.round(samples / self._stepsizes) * self._stepsizes + 0
```

Comment:
Mightn't this produce values outside the box? E.g. wouldn't `OrdinalSearchSpace([0.0], [1.0], [0.6])` round a 0.95 sample to 1.2?

(If that's a problem then do we need to make sure that the solution doesn't skew the sampling distribution? E.g. if we just round 0.95 down to 0.6 then that'd make 0.6 more likely than 0.0. Perhaps we should initialise lower and upper to be the smallest/biggest multiples of stepsize in the box, and document that in the docstring?)

Reply:
Yeah it produce values outside the box, my bad.

> Perhaps we should initialise lower and upper to be the smallest/biggest multiples of stepsize in the box, and document that in the docstring?)

do you mean we should replace the user predefined bound to that smallest/biggest multiples? 
what if we just filter out of bound point and replace it with lower/upper bound? this will make the box not "evenly spaced" again though.

Reply:
Yes, my suggestion was changing the bounds so that they're multiples. @henrymoss Do you have an opinion on this?

Reply:
Yep . Lets fix stepsizes to be multiples!

Reply:
added, please check. however, it seems need more test points inside `test_ordinalsearchspace_bounds_changes_to_nearest_multiples` , since there might be some floating-point problems.